### PR TITLE
python312Packages.soundcard init at 0.4.3

### PR DESCRIPTION
**Pull request by matthiasdotsh**
## Description of changes

https://github.com/bastibe/SoundCard

A python package to interact with your soundcard.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [X] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/soundcard/default.nix
Comment (TomaSajt) [inline pkgs/development/python-modules/soundcard/default.nix:20]: IIRC `pname` is no longer allowed to contain uppercase characters. This means you can no longer do `inherit pname` in `fetchPypi`.

```suggestion
  pname = "soundcard";
  version = "0.4.3";
  pyproject = true;

  src = fetchPypi {
    pname = "SoundCard";
    inherit version;
    hash = "sha256-QQg1UUuhCAmAPLmIfUJw85K1nq82WRW7lFFq8/ix0Dc=";
  };
```

Comment (TomaSajt) [reply pkgs/development/python-modules/soundcard/default.nix:20]: Though it might also be better to use `fetchFromGitHub` instead of `fetchPypi` since `fetchFromGitHub` is a more easily inspectable source.

Comment (matthiasdotsh) [reply pkgs/development/python-modules/soundcard/default.nix]: Thanks for the comment. I was not aware about the `pname` thing. I just pushed a version with all lowercase `pname` and `fetchFromGitHub`.

Anything else I should change before this can be merged?

**Comment by matthiasdotsh** [comment]
fore-push: update `pname` to all lowercase + use `fetchFromGitHub`

File: pkgs/development/python-modules/soundcard/default.nix
Comment (TomaSajt) [inline pkgs/development/python-modules/soundcard/default.nix:27]: The newer `--replace-fail` flags should be used instead of the legacy `--replace` flag, since it won't silently fail.

Also, I think it's safer to also include `dlopen` in the replacement to not accidentally replace other instances of the `'pulse'` literal.

And finally, `libpulse.so` is only valid on linux systems. On darwin/Mac it's not `.so` but `.dylib`. `stdenv.hostPlatform.extensions.sharedLibrary` can be used instead, which evaluates to `.so` or `.dylib` depending on which system you're building for.


```suggestion
  patchPhase =
    let
      libpulse = "${libpulseaudio}/lib/libpulse${stdenv.hostPlatform.extensions.sharedLibrary}";
    in
    ''
      substituteInPlace soundcard/pulseaudio.py \
          --replace-fail "_ffi.dlopen('pulse')" "_ffi.dlopen('${libpulse}')"
    '';
```

File: pkgs/development/python-modules/soundcard/default.nix
Comment (TomaSajt) [inline pkgs/development/python-modules/soundcard/default.nix:19]: This is a bit of a nit, but nowadays it's better practice not to use `pname` for substituting occurrences of the package name.

```suggestion
    repo = "soundcard";
```

**Review by TomaSajt — CHANGES_REQUESTED** [review]
The commit message (and PR title) should be the following:
`python312Packages.soundcard: init at 0.4.3`

This is needed because the automatic build system CI won't use aliases. (`pyton3Packages` is an alias of `python311Packages`)

File: pkgs/development/python-modules/soundcard/default.nix
Comment (matthiasdotsh) [reply pkgs/development/python-modules/soundcard/default.nix:27]: Thank you very much for your advice here.
I don't have a MacOS to test right now, so I'm super grateful for feedback.

The [Upstream project](https://github.com/bastibe/SoundCard) says.

> SoundCard is a library for playing and recording audio without resorting to a CPython extension. Instead, it is implemented using the wonderful [CFFI](http://cffi.readthedocs.io/en/latest/) and the native audio libraries of Linux, Windows and macOS.

> SoundCard is cross-platform, and supports Linux/pulseaudio, Mac/coreaudio, and Windows/WASAPI. 


I'm not sure whether `pulseaudio` or `coraudio` should be used on a macOS with nix as package manager.

Comment (TomaSajt) [reply pkgs/development/python-modules/soundcard/default.nix]: I'll try to look into it and try it out

**Comment by TomaSajt** [comment]
The `python312Packages` name is case-sensitive, so please update it to use lowercase starting `p`

File: pkgs/development/python-modules/soundcard/default.nix
Comment (TomaSajt) [reply pkgs/development/python-modules/soundcard/default.nix]: Okay, I asked around on the Matrix chat about `dlopen` on darwin, still waiting for a response.

The current patching should be changed, then, to only happen on linux.
This can be done with
```nix
postPatch = lib.optionalString stdenv.isLinux ''
  substituteInPlace ....
'';
```

File: pkgs/development/python-modules/soundcard/default.nix
Comment (TomaSajt) [inline pkgs/development/python-modules/soundcard/default.nix:24]: This might seem a bit weird, but `postPatch` should be used instead of `patchPhase`, since if you override the default logic of `patchPhase`, the `patches = [ ... ];` list will not work, if you ever want to use `.patch` files for this package in the future.

```suggestion
  postPatch = ''
```

File: pkgs/development/python-modules/soundcard/default.nix
Comment (matthiasdotsh) [reply pkgs/development/python-modules/soundcard/default.nix]: I just pushed a variant with `replace-fail`, where the patch is only applied (in `postPatch`) when we are on a linux system.
In this case I assume it's fine to use `.so` as extension.

**Review by TomaSajt — APPROVED** [review]
The package itself LGTM

I don't have experience writing/reading NixOS tests, so I am not able to provide much feedback for that.

---

A final nit would be the ordering of the top attrset:
I'd really recommend putting the python dependencies next to each other and putting `lib` above everything.

---

_In general, this is my personal preference, which you don't have to follow:
`lib`, `stdenv`, the builder-function, fetchers, hooks, build-time deps, normal deps, test deps, other._
_I use this because usually this is the order they appear in the derivation._

**Comment by matthiasdotsh** [comment]
> A final nit would be the ordering of the top attrset:
> I'd really recommend putting the python dependencies next to each other and putting lib above everything.

I have just sorted the dependencies.
The way you do it sounds good to me at first. I'm still new to the nix environment and would be grateful for any feedback.

I've learned a lot from your review, thank you very much!


> The package itself LGTM

What happens now with the MR? Can you merge it directly? Is there anything else I need to do?

**Comment by TomaSajt** [comment]
Well, I don't have any commit access to the repo, so you'll probably have to wait for someone who does to comes along.

The waiting time can be improved a bit by submitting this PR into some other place, such as
the "PRs already reviewed" thread on the NixOS discourse (https://discourse.nixos.org/t/prs-already-reviewed/2617/1599)
or maybe the Matrix chat for Review Requests (https://matrix.to/#/#review-requests:nixos.org)

I'll probably post it to "PRs already reviewed" after I do some tests on a Mac emulator later today.

File: pkgs/development/python-modules/soundcard/test.nix
Comment (TomaSajt) [reply pkgs/development/python-modules/soundcard/test.nix:29]: Hmm, interesting...
Could you share the resources about how to run this?

Comment (matthiasdotsh) [reply pkgs/development/python-modules/soundcard/test.nix:29]: Sure: https://nixcademy.com/2023/12/01/nixos-integration-tests-part-2/

**Comment by matthiasdotsh** [comment]
Force push: rebase to latest master and fix merge conflict in `maintainers/maintainer-list.nix`

File: pkgs/development/python-modules/soundcard/default.nix
Comment (TomaSajt) [inline pkgs/development/python-modules/soundcard/default.nix:13]:
```suggestion
  pname = "soundcard";
```

File: pkgs/development/python-modules/soundcard/default.nix
Comment (TomaSajt) [inline pkgs/development/python-modules/soundcard/default.nix:18]:
```suggestion
    pname = "SoundCard";
    inherit version;
```

since `pname` is no longer the same as the pypi name, we have to manually provide the name

File: pkgs/development/python-modules/soundcard/default.nix
Comment (TomaSajt) [reply pkgs/development/python-modules/soundcard/default.nix:12]: still, the the `pname` that is provided for `buildPythonPackage` should be lowercase.

only the `fetchPypi`'s `pname` should have any upper case letters.

You can keep `version` in the `let in` block if you  don't want to use `rec`, but I think `pname` should just stay inside `buildPythonPackage`, since it's not needed anywhere else.

**Review by Sigmanificient — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/soundcard/default.nix
Comment (matthiasdotsh) [reply pkgs/development/python-modules/soundcard/default.nix:44]: it is copied from the [upstream project](https://github.com/bastibe/SoundCard) ...

File: pkgs/development/python-modules/soundcard/default.nix
Comment (TomaSajt) [reply pkgs/development/python-modules/soundcard/default.nix:46]: It might be better to change it to
`"https://github.com/bastibe/SoundCard/blob/${version}/README.rst#changelog"`
since we can use the current version this way.

File: pkgs/development/python-modules/soundcard/test.nix
Comment (matthiasdotsh) [reply pkgs/development/python-modules/soundcard/test.nix]: You can run the test with `nix-build -A pkgs.python3Packages.soundcard.passthru.tests.vm-with-soundcard`

File: pkgs/development/python-modules/soundcard/default.nix
Comment (matthiasdotsh) [reply pkgs/development/python-modules/soundcard/default.nix:46]: do you have a suggestion? it looks to me like there are no tags available?

File: pkgs/development/python-modules/soundcard/default.nix
Comment (Sigmanificient) [reply pkgs/development/python-modules/soundcard/default.nix:44]: See https://github.com/NixOS/nixpkgs/tree/master/pkgs#meta-attributes

`meta.description` must:

- Be short, just one sentence.
- Be capitalized.
- Not start with the definite or an indefinite article.
- Not start with the package name.
  -  More generally, it should not refer to the package name.
- Not end with a period (or any punctuation for that matter).
- Provide factual information.
  - Avoid subjective language.

Comment (matthiasdotsh) [reply pkgs/development/python-modules/soundcard/default.nix]: Thank you, I wasn't aware of that.

**Pull request merged**
